packages/expo-device/ios/EXDevice/EXDevice.m

```diff
@@ -0,0 +1,252 @@
+                    resolver:(UMPromiseResolveBlock)resolve
+                    rejecter:(UMPromiseRejectBlock)reject) {
+  //some iOS privacy issues
+  NSString *address = @"02:00:00:00:00:00";
+  resolve(address);
+}
+
+UM_EXPORT_METHOD_AS(getIPAddressAsync,
+                    getIPAddressAsyncWithResolver:(UMPromiseResolveBlock)resolve rejecter:(UMPromiseRejectBlock)reject)
+{
+  NSString *address = @"0.0.0.0";
+  struct ifaddrs *interfaces = NULL;
+  struct ifaddrs *temp_addr = NULL;
+  int success = 0;
+  // retrieve the current interfaces - returns 0 on success
+  success = getifaddrs(&interfaces);
+  if (success == 0) {
+    // Loop through linked list of interfaces
+    temp_addr = interfaces;
+    while(temp_addr != NULL) {
+      if(temp_addr->ifa_addr->sa_family == AF_INET) {
+        // Check if interface is en0 which is the wifi connection on the iPhone
+        if([[NSString stringWithUTF8String:temp_addr->ifa_name] isEqualToString:@"en0"]) {
+          // Get NSString from C String
+          address = [NSString stringWithUTF8String:inet_ntoa(((struct sockaddr_in *)temp_addr->ifa_addr)->sin_addr)];
+        }
+      }
+      temp_addr = temp_addr->ifa_next;
+    }
+  }
+  // Free memory
+  freeifaddrs(interfaces);
+  resolve(address);
+}
+
+UM_EXPORT_METHOD_AS(isPinOrFingerprintSetAsync, isPinOrFingerprintSetAsyncWithResolver:(UMPromiseResolveBlock)resolve rejecter:(UMPromiseRejectBlock)reject)
+{
+#if TARGET_OS_TV
+  BOOL isPinOrFingerprintSet = false;
+#else
+  LAContext *context = [[LAContext alloc] init];
+  BOOL isPinOrFingerprintSet = ([context canEvaluatePolicy:LAPolicyDeviceOwnerAuthentication error:nil]);
+#endif
+  resolve(@[[NSNumber numberWithBool:isPinOrFingerprintSet]]);
+}
+
+UM_EXPORT_METHOD_AS(getFreeDiskStorageAsync, getFreeDiskStorageAsyncWithResolver:(UMPromiseResolveBlock)resolve rejector:(UMPromiseRejectBlock)reject)
+{
+  NSString *returnString = [NSString stringWithFormat:@"%llu", self.freeDiskStorage];
+  resolve(returnString);
+}
+
+- (NSString *)deviceId
+{
+  struct utsname systemInfo;
+  
+  uname(&systemInfo);
+  
+  NSString* deviceId = [NSString stringWithCString:systemInfo.machine
+                                          encoding:NSUTF8StringEncoding];
+  
+  if ([deviceId isEqualToString:@"i386"] || [deviceId isEqualToString:@"x86_64"] ) {
+    deviceId = [NSString stringWithFormat:@"%s", getenv("SIMULATOR_MODEL_IDENTIFIER")];
+  }
+  
+  return deviceId;
+}
+
+- (NSString *)carrier
+{
+#if (TARGET_OS_TV)
+  return nil;
+#else
+  CTTelephonyNetworkInfo *netinfo = [[CTTelephonyNetworkInfo alloc] init];
+  CTCarrier *carrier = [netinfo subscriberCellularProvider];
+  return carrier.carrierName;
+#endif
+}
+
+/*device types*/
+typedef NS_ENUM(NSInteger, DeviceType) {
+  DeviceTypeHandset,
+  DeviceTypeTablet,
+  DeviceTypeTv,
+  DeviceTypeUnknown
+};
+
+#define DeviceTypeValues [NSArray arrayWithObjects: @"Handset", @"Tablet", @"Tv", @"Unknown", nil]
+
+- (DeviceType)getDeviceType
```

Comment:
Why don't we simplify handling device type and instead of defining, maintaining our own enum:
- create a method that would convert `UIUserInterfaceIdiom` to one of `Handset`, `Tablet`, `Tv`, `Unknown`
    `- (NSString *)deviceType`
- return its value in `constantsToExport`
- in `isTablet` method, which imho could be removed in favor of a simple `@"isTablet": @([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad),` compare interface idiom (not `DeviceType`)
- remove `DeviceTypeValues` macro and enum

Reply:
Thanks for the suggestions @sjchmiela! They are reflected in the updated commits. One question: 
> * in `isTablet` method, which imho could be removed in favor of a simple `@"isTablet": @([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad),` compare interface idiom (not `DeviceType`)
After I tried implement that change, `isTablet` returned `0` when I tested. This is probably because in Objective-C, the `==` operator implicitly returns `0` or `1` like in C, so I decided to put expression in the `- (bool) isTablet` method.